src/stan/variational/advi.hpp

```diff
@@ -150,18 +150,143 @@ namespace stan {
+          eta_sequence.pop();
+
+          for (iter_tune = 1; iter_tune <= tuning_iter; ++iter_tune) {
+            m = (eta_sequence_size - eta_sequence.size() - 1) *
+              tuning_iter + iter_tune; // # of total tuning iterations
+            stan::services::variational::print_progress(
+              m, 0, tuning_iter*eta_sequence_size,
+              tuning_iter, true, "", "", *print_stream_);
+
+            // Compute gradient of ELBO
+            calc_ELBO_grad(variational, elbo_grad);
+
+            // Update learning rate parameters
+            if (iter_tune == 1) {
+              params_prop += elbo_grad.square();
+            } else {
+              params_prop = pre_factor * params_prop +
+                            post_factor * elbo_grad.square();
+            }
+            eta_scaled = eta / sqrt(static_cast<double>(iter_tune));
+
+            // Stochastic gradient update
+            variational += eta_scaled * elbo_grad / (tau + params_prop.sqrt());
+          }
+          elbo = calc_ELBO(variational);
+
+          // Check if:
+          // (1) ELBO at current eta is worse than the best ELBO
+          // (2) the best ELBO hasn't actually diverged
+          if (elbo < elbo_best && elbo_best > elbo_init) {
+            if (print_stream_) {
+              *print_stream_ << "Success!"
+                << " Found best tuned hyperparameter"
+                << " [eta = " << eta_best
+                << "]";
+              if (eta_sequence.size() > 0) {
+                *print_stream_
+                  << " earlier than expected.";
+              }
+              *print_stream_ << std::endl << std::endl;
+            }
+            do_more_tuning = false;
+          } else {
+            if (eta_sequence.size() > 0) {
+              // Restart phase
+              elbo_best = elbo;
+              eta_best = eta;
+            } else {
+              // No more eta values to try, so use current eta if it
+              // didn't diverge or fail if it did diverge
+              if (elbo > elbo_init) {
+                if (print_stream_)
+                  *print_stream_ << "Success!"
+                    << " Found best tuned hyperparameter"
+                    << " [eta = " << eta_best
+                    << "]" << std::endl << std::endl;
+                eta_best = eta;
+                do_more_tuning = false;
+              } else {
+                const char* name = "All proposed step-sizes";
+                const char* msg1 = "failed. Your model may be either "
+                  "severely ill-conditioned or misspecified.";
+                stan::math::domain_error(function, name, "", msg1);
+                return 0;
+              }
+            }
+            // Reset
+            params_prop = Q(model_.num_params_r());
+          }
+          variational = Q(cont_params_);
+        }
+        return eta_best;
+      }
+
+      /**
+       * Runs stochastic gradient ascent with an adaptive stepsize.
```

Comment:
Also, don't be shy about replacing this

```
bool flag = true;
while (flag) {
  if (cond) {
    ...
    flag = false;
  } else {
    ...
    flag = true;
  }
}
```

with

```
while (true) {
  if (cond) {
     ...
     break;
  } else {
     ...
  }
}
```

I find code with flags very hard to follow because you need to track 
through the whole function to figure out where and how the flag gets reset.
Here, I think there's probably also needless call to the constructor
at the end when the flag is set to false before exit.

Also, because params_prop is only used in that loop, you should declare
it close to the loop.  The general style rule is to declare variables as
close to where they are used as possible.  

Also, this sequence:

```

        eta_sequence.push(1.00);
        eta_sequence.push(0.50);  // 1/2 of previous value
        eta_sequence.push(0.10);  // 1/5 of previous value
        eta_sequence.push(0.05);  // 1/2 of previous
        eta_sequence.push(0.01);  // 1/5 of previous
```

isn't evenly spaced on the log scale, so I have no idea why you
chose those numbers other than they match American currency denominations!
(So I'm missing the 0.25!).  


There's also this

```

        double eta;
        double eta_best;
```

which is bad for several reasons.  There are no initial values and
it's not declared by where it's used and it's not used in the scope
where it's declared.  Instead, you should replace this

```

          eta = eta_sequence.front();
```

with

```
          double eta = eta_sequence.front();
```

Also, I don't like that whole loop.  You should just iterate through
eta values declared as a static const array, or as a vector if you must
rather than popping a sequence.  It makes it clearer what happens when you run 
out of values.  So this:

```
        while (do_more_tuning) {
          // Try next eta
          eta = eta_sequence.front();
          eta_sequence.pop();
```

should be rewritten as

```
  for (size_t i = 0; i < eta_sequence.size(); ++i)
    eta = eta_sequence[i];
```

or use iterators if you're comfortable with that idiom.  I'm not clear
what the intended behavior is in the current code for when you pop the last
value off the eta_sequence.

- Bob

> On Aug 26, 2015, at 8:57 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> > +                const char* msg1 = "failed. Your model may be either "
> > +                  "severely ill-conditioned or misspecified.";
> > +                stan::math::domain_error(function, name, "", msg1);
> > +                return 0;
> > +              }
> > +            }
> > +            // Reset
> > +            params_prop = Q(model_.num_params_r());
> > +          }
> > +          variational = Q(cont_params_);
> > +        }
> > +        return eta_best;
> > +      }
> > +
> > +      /**
> > +       * Runs stochastic gradient ascent with an adaptive stepsize.
> 
> can you describe the input and output more? it's hard to tell that the thing that this function is updating is the variational parameter. What's expected to happen at the end of this function?
> 
> —
> Reply to this email directly or view it on GitHub.
>